test/params_test.c

```diff
@@ -0,0 +1,549 @@
+static size_t bignumbin_l;            /* "p3" */
+static char app_p4[256];              /* "p4" */
+static size_t app_p4_l;               /* "p4" */
+static char app_p5[256];              /* "p5" */
+static size_t app_p5_l;               /* "p5" */
+static const char *app_p6 = NULL;     /* "p6" */
+static size_t app_p6_l;               /* "p6" */
+static unsigned char foo[1];          /* "foo" */
+static size_t foo_l;                  /* "foo" */
+
+#define app_p1_init 17           /* A random number */
+#define app_p2_init 4711.0       /* Another random number */
+#define app_p3_init "deadbeef"   /* Classic */
+#define app_p4_init "Hello"
+#define app_p5_init "World"
+#define app_p6_init "Cookie"
+#define app_foo_init 'z'
+
+static int cleanup_app_variables(void)
+{
+    BN_free(app_p3);
+    app_p3 = NULL;
+    return 1;
+}
+
+static int init_app_variables(void)
+{
+    int l = 0;
+
+    cleanup_app_variables();
+
+    app_p1 = app_p1_init;
+    app_p2 = app_p2_init;
+    if (!BN_hex2bn(&app_p3, app_p3_init)
+        || (l = BN_bn2nativepad(app_p3, bignumbin, sizeof(bignumbin))) < 0)
+        return 0;
+    bignumbin_l = (size_t)l;
+    strcpy(app_p4, app_p4_init);
+    app_p4_l = sizeof(app_p4_init);
+    strcpy(app_p5, app_p5_init);
+    app_p5_l = sizeof(app_p5_init);
+    app_p6 = app_p6_init;
+    foo[0] = app_foo_init;
+    foo_l = sizeof(app_foo_init);
+
+    return 1;
+}
+
+/*
+ * Here, we define test OSSL_PARAM arrays
+ */
+
+/* An array of OSSL_PARAM, specific in the most raw manner possible */
+static const OSSL_PARAM static_raw_params[] = {
+    { "p1", OSSL_PARAM_INTEGER, &app_p1, sizeof(app_p1), NULL },
+    { "p3", OSSL_PARAM_UNSIGNED_INTEGER, &bignumbin, sizeof(bignumbin),
+      &bignumbin_l },
+    { "p4", OSSL_PARAM_UTF8_STRING, &app_p4, sizeof(app_p4), &app_p4_l },
+    { "p5", OSSL_PARAM_UTF8_STRING, &app_p5, sizeof(app_p5), &app_p5_l },
+    { "p6", OSSL_PARAM_UTF8_PTR, &app_p6, sizeof(app_p6), &app_p6_l },
+    { "foo", OSSL_PARAM_OCTET_STRING, &foo, sizeof(foo), &foo_l },
+    { NULL, 0, NULL, 0, NULL }
+};
+
+/* The same array of OSSL_PARAM, specified with the macros from params.h */
+static const OSSL_PARAM static_api_params[] = {
+    OSSL_PARAM_int("p1", &app_p1),
+    OSSL_PARAM_SIZED_BN("p3", &bignumbin, sizeof(bignumbin), bignumbin_l),
+    OSSL_PARAM_DEFN("p4", OSSL_PARAM_UTF8_STRING,
+                    &app_p4, sizeof(app_p4), &app_p4_l),
+    OSSL_PARAM_DEFN("p5", OSSL_PARAM_UTF8_STRING,
+                    &app_p5, sizeof(app_p5), &app_p5_l),
+    OSSL_PARAM_DEFN("p6", OSSL_PARAM_UTF8_PTR,
+                    &app_p6, sizeof(app_p6), &app_p6_l),
+    OSSL_PARAM_DEFN("foo", OSSL_PARAM_OCTET_STRING, &foo, sizeof(foo), &foo_l),
+    OSSL_PARAM_END
+};
+
+/*
+ * The same array again, but constructed at run-time
+ * This exercises the OSSL_PARAM constructor functions
+ */
+OSSL_PARAM *construct_api_params(void)
+{
+    size_t n = 0;
+    static OSSL_PARAM params[10];
+    OSSL_PARAM param_end = OSSL_PARAM_END;
```

Comment:
Might the param end variable make more sense as a global static const? Which would avoid this local everywhere it is needed. Even a function call that returned the structure would be cleaner.

Reply:
Why not both?  `OSSL_PARAM_end` being the global constant and `OSSL_PARAM_END` being the macro...

Reply:
Channeling for @dot-asm:  it's not good (bad, actually) to have syntactic sugar for C.  ```{NULLL}``` suffices for END; use that.